### ISSUE TITLE
Transforms for get, assoc, update & *-in forms

### ISSUE THREAD

**Issue by snoe** [clojure-lsp/clojure-lsp#904]
**Is your feature request related to a problem? Please describe.**
Often things change and what used to be deeply nested is at the root, and vice versa. Manually changing code from one form to the other can be cumbersome. Maybe `transform-path-fn [tp]` for a name? It would cycke back and forth between forms.

Simplest
`(get m k)` <=> `(get-in m [k])`
`(assoc m k v)` <=> `(assoc-in m [k] v)`
`(update m k f f-args)` <=> `(update-in m [k] f f-args)`

If `k` is a keyword
`(:k m)` <=> `(get-in m [:k])`

Potentially support multiple gets
`(:k (:j m))` <=> `(get-in m [:j :k])`

Potentially work in thread-first
`(-> m :j :k inc)` <=> `(-> m (get-in [:j :k) inc)`
`(-> m :j :k)` <=> `(get-in m [:j :k])`

**Potential other solutions**
Instead of a cycling transform, it could be more like `thread` (one/all) and `unwind` (one/all). But I don't think I've ever wanted intermediate steps like this.

`(:k (:j m))` =in-one> `(:k (get-in m [:j])` =in-one> `(get-in m [:j :k])`
`(:k (:j m))` =in-all> `(get-in m [:j :k])`
`(get-in m [:j :k])` =?-all>`(:k (:j m))`

**Comment by ericdallo** [comment]
Sounds good. the refactor name is a little bit weird to me, maybe we are trying to fit mnemonics and making the name more confusing that it could be?
Since this is a `cycle` operation, it could be cycle-path or something? the path name sounds odd still to me

I like the intermediate since we can leave to users decide if they want to stop or keep applying, while the "apply all" version doesn't support that.

**Comment by snoe** [comment]
> Sounds good. the refactor name is a little bit weird to me, maybe we are trying to fit mnemonics and making the name more confusing that it could be? Since this is a `cycle` operation, it could be cycle-path or something? the path name sounds odd still to me

Yes, unfortunately `cp` is taken, and that's the tradeoff, gotta come up with unique mnemonics (the code action can say whatever).

> I like the intermediate since we can leave to users decide if they want to stop or keep applying, while the "apply all" version doesn't support that.

When would you use the intermediate? I couldn't think of a use case. If there's only an intermediate, I also worry that it would become cumbersome like `expand-let` and applying that over and over.

**Comment by ericdallo** [comment]
> Yes, unfortunately cp is taken, and that's the tradeoff, gotta come up with unique mnemonics (the code action can say whatever).

Yeah, I suggest we stick with `cycle` and find a better name for `path` maybe

> When would you use the intermediate? I couldn't think of a use case. If there's only an intermediate, I also worry that it would become cumbersome like expand-let and applying that over and over.

Yeah, I think so, np following with the "all", and we could re-think if someone complains what I doubt it will happen soon

**Comment by mainej** [comment]
I agree, occasionally I want to convert between `get/assoc/update` and `get-in/assoc-in/update-in`. 

But much more commonly I want to convert between `(:x m)` and `(get m :x)` or between `(:y (:x m))` and `(get-in m [:x :y])`.

Should `no-in <=> in`  and `parens <=> get` be two different features?

Actually the aspect that interests me the most is an "unwinding" variation that hasn't been mentioned yet: `(get-in m [:x :y :z]) => (get-in (:x m) [:y :z])`. This is useful if you want to compose these transformations with destructuring `m`'s keys. I do this especially often when I have code like this:

```clojure
(defn foo [m]
  (* (get-in m [:x :y :z])
     (:n (:x m))))
```

I notice that this code accesses `:x` in several different ways, and that it'd be clearer if `:x` were destructured once in the defn binding, to emphasize that the code is about `x`, not about `m`:

```clojure
;; unwind
(defn foo [m]
  (* (get-in (:x m) [:y :z])
     (:n (:x m))))

;; destructure
(defn foo [{:keys [x]}]
  (* (get-in x [:y :z])
     (:n x)))
```

Being able to unwind `get` in the way I described makes it easier to coerce everything into the same form `(:x m)`. Then clj-refactor or #1139 have to understand only one form: `(:x m) -> x`.

Comparing that to the original examples it becomes clear there are two possibilities for `(get-in m [:x :y :z])`. You could want to unwind the last element in the get path `(:z (get-in m [:x :y]))`, or the first element in the get path `(get-in (:x m) [:y :z])`. As a longer example, there are two valid routes to get between the same start and end points:

```clojure
;; unwind last in get path

|(get-in m [:x :y :z])
(:z |(get-in m [:x :y]))
;; (:z (:y |(get-in m [:x]))) ;; if get-in -> get should be explicit instead of implied
(:z (:y |(get m :x)))
(:z (:y (:x |m)))

;; unwind first in get path

|(get-in m [:x :y :z])
|(get-in (:x m) [:y :z])
;; |(get-in (:y (:x m)) [:z])
|(get (:y (:x m)) :z)
|(:z (:y (:x m)))
```

I think both have their uses. So here's a more complete proposal:

**`get-in-all`**: Like "thread all", there's a command that does "as much getting-in as possible, working down". It's invoked from outside a form.
```clojure
|(:z (:y (:x m)))
;; get-in-all
|(get-in m [:x :y :z])

;; but

(:z |(:y (:x m)))
;; get-in-all
(:z |(get-in m [:x :y]))
```
_Notes:_
1. I'm using keywords, but I think this command should be agnostic about the contents of the form.
2. What if the form isn't a series of 1-arg calls? I think we'd give up if we reached a 3-arg call, but a 2-arg call could be providing a default: `|(:x m "default")` -> `|(get m :x "default)`. These defaults don't nest nicely. Something to think about.

**`get-in-unwind-all`** Like the threading version of "unwind all", there'd be a command that is the inverse of `get-in-all` that removes as much "getting" as possible. As with the threading version this command could prefer to unwind a `get/get-in` under the cursor, but look "up" as a fallback.

**`unwind-first-in-get-path`** and **`unwind-last-in-get-path`** To match "unwind one" there'd be a pair of commands to unwind one level of a `get-in`. As defined above, they would unwind from different ends of the get-in path.

**`get-in-one`** Finally I can imagine a tool to add one level of "getting", though as discussed earlier, it could be added later. It might also need two variations, one where it adds to the front of the path and one where it adds to the back of the path.

And as a final tangential comment. I think it'd be nicer both as a `clojure-lsp` user and as a contributor if this feature were orthogonal to threading.  That is, rather than the proposed
```clojure
(-> m :j :k inc)
(-> m (get-in [:j :k]) inc)
```

I think it's more powerful and flexible if you accomplish the same thing by composing refactorings, perhaps:

```clojure
(-> m :j :k inc) ;; unwind-all ->
(inc (:k (:j m))) ;; get-in-all -> 
(inc (get-in m [:j :k])) ;; thread-first-all ->
(-> m (get-in [:j :k]) inc)
```

**Comment by mainej** [comment]
@snoe I've been thinking about this more. I feel like I have a coherent proposal about "threading" `get`s. But I also feel like adding those details here would be hijacking your thread. This issue, as its title suggests, is about converting `update` to `update-in`, and similarly for `assoc` and `get`. Those conversions have a little overlap with threading, but I think they're really separate things. So, I'm going to start a new issue. After it's written, I'll link to it here.

**Comment by mainej** [comment]
See https://github.com/clojure-lsp/clojure-lsp/issues/1172 for an issue focused on threading `get`.
